### PR TITLE
add `anvil_dumpState` and `anvil_loadState`

### PR DESCRIPTION
**Pull request by dbeal-eth**
## Motivation

[Cannon](https://usecannon.com/) is a packaging and deployment system for EVM protocols. One of the core tenants of its operation requires saving snapshots of the state of a chain (similar to how a [Docker](https://www.docker.com/) image saves snapshots of fs layers) and using them to transmit and recreate in a new EVM dev instance. It is not convenient, reliable, or performant to get the chain state through individual RPC calls. Therefore, a dedicated command is necessary.

## Solution

Create new commands `anvil_dumpState` and `anvil_loadState`.

* `anvil_dumpState` takes no arguments. Upon invocation, returns a large hex string which can be loaded into a separate/restarted anvil instance in order to reattain the entire chain state. Only supported on standalone (ie non-fork) networks.
* `anvil_loadState` takes one argument, a hex string previously dumped with `anvil_dumpState`. Upon invocation, merges the specified state into the currently active network. **The merge functionality is very important. Cannon will regularly clear the state using `evm_revert` and load a sequence of states of separate contracts. States should also be merged within the same account.** It is the responsibility of the RPC client to verify there are no storage collisions.

## Limitations

I originally wanted to have the returned data compressed, but no compression library is included in any of foundry's packages. If you are OK with a compression package being added to the cargo dependencies, please LMK and the data will be deflated before being returned or something.

I have not done any testing to verify what the maximum size the RPC response is. However, I have run tests with blobs that ended up being ~5MB and there were no problems. If you happen to know the answer to this question (for example if log calls) please share.

## Additional Info

There is a companion PR open for the same feature on Hardhat network https://github.com/NomicFoundation/hardhat/pull/2425 . @alcuadrado has confirmed this will be supported after they are done with some refactoring on their end. The hardhat PR will be updated by me to match the standard set by this PR.

Am I supposed to add tests for this? If so, please LMK where the recommended place to add the tests is.

### CODE REVIEW

**Review by onbjerg — COMMENTED** [review]
Code-wise looks good to me, I don't really have an opinion on the feature itself so I will defer to @gakonst and @mattsse on that. Going by the code here, it seems that Hardhat has this functionality? If so, seems OK to include it in Anvil, too.

**Comment by dbeal-eth** [comment]
> It would be cool if this was compatible with geth's `debug_dumpBlock` - https://github.com/ethereum/go-ethereum/blob/2697e44d819377e39a781e5ab9f1814426b4b0f0/eth/api.go#L258
> 
> It looks like its not 100% the same, but pretty close, this is what is returned by geth - https://github.com/ethereum/go-ethereum/blob/2697e44d819377e39a781e5ab9f1814426b4b0f0/core/state/dump.go#L64

didnt know this function existed, but it is convenient that the formats are so close to the same. Definitely can make the few changes to match it up :+1: 

Does geth have a corresponding `load` function as well, or is it just the dump for debugging?

File: anvil/src/eth/backend/mem/mod.rs [hunk 1]
```diff
@@ -407,6 +407,37 @@ impl Backend {
...
+                    "Dumping state not supported with the current configuration",
+                )
+                .into()
+            })
+    }
+
+    /// Deserialize and add all chain data to the backend storage
+    pub fn load_state(&self, buf: Bytes) -> Result<bool, BlockchainError> {
+        let state: SerializableState = serde_json::from_str(
+            str::from_utf8(&buf.0).map_err(|_| BlockchainError::FailedToDecodeStateDump)?,
```
Comment (mattsse) [inline anvil/src/eth/backend/mem/mod.rs:427]: `serde_json::from_slice()` should be able to deserialize this directly

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for this, 

overall this looks great. only a couple of smol style nits.

not familiar with `debug_dumpBlock` so unsure whether it's possible to unify?

File: anvil/src/eth/backend/mem/mod.rs (hunk 1 above)
Comment (dbeal-eth) [reply anvil/src/eth/backend/mem/mod.rs]: 2259dd8

File: anvil/src/eth/backend/mem/mod.rs [hunk 2]
```diff
@@ -407,6 +407,37 @@ impl Backend {
         self.db.write().revert(id)
     }
 
+    /// Write all chain data to serialized bytes buffer
+    pub fn dump_state(&self) -> Result<Bytes, BlockchainError> {
+        self.db
+            .read()
+            .dump_state()
+            .map(|s| bytes::Bytes::from(serde_json::to_string(&s).unwrap_or_default()).into())
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply anvil/src/eth/backend/mem/mod.rs]: 2259dd8

File: anvil/src/eth/backend/mem/in_memory_db.rs [hunk 3]
```diff
@@ -20,6 +22,59 @@ impl Db for MemDb {
...
+                                .info
+                                .code
+                                .unwrap_or_else(|| self.inner.code_by_hash(v.info.code_hash))
+                                .into(),
+                            storage: v.storage.into_iter().collect(),
+                        },
+                    )
+                })
+                .collect(),
+        })
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply anvil/src/eth/backend/mem/in_memory_db.rs]: 2259dd8

File: anvil/src/eth/backend/db.rs [hunk 4]
```diff
@@ -10,8 +10,24 @@ use forge::revm::KECCAK_EMPTY;
...
+    pub accounts: HashMap<Address, SerializableAccountRecord>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct SerializableAccountRecord {
+    pub nonce: u64,
+    pub balance: U256,
+    pub code: Bytes,
+    pub storage: HashMap<U256, U256>,
+}
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply anvil/src/eth/backend/db.rs]: 2259dd8

File: anvil/src/eth/backend/db.rs [hunk 5]
```diff
@@ -118,3 +148,15 @@ impl DatabaseRef for StateDb {
         self.0.block_hash(number)
     }
 }
+
+impl SerializableState {
+    pub fn new() -> SerializableState {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply anvil/src/eth/backend/db.rs]: 2259dd8

**Comment by tynes** [comment]
> so I g uess my concern is that the dump is kind of useless if it only has the code_hash instead of the actual code

The example I showed above was the result of `debug_dumpBlock` for an account with no code, the `code` key isn't serialized in the json. When an account does have code, it will have a `code` key, see here: https://github.com/ethereum/go-ethereum/blob/2697e44d819377e39a781e5ab9f1814426b4b0f0/core/state/dump.go#L56

**Comment by dbeal-eth** [comment]
> > so I g uess my concern is that the dump is kind of useless if it only has the code_hash instead of the actual code
> 
> The example I showed above was the result of `debug_dumpBlock` for an account with no code, the `code` key isn't serialized in the json. When an account does have code, it will have a `code` key, see here: https://github.com/ethereum/go-ethereum/blob/2697e44d819377e39a781e5ab9f1814426b4b0f0/core/state/dump.go#L56

I see now that the code hash there is the hash of the empty buffer, which makes sense now. So the code of a contract can be guaranteed to be set if it is has code, and I assume same goes for storage.

Do you happen to have a small sample dump handy? If not, I will try and get a dump from my local eth node tomorrow or something

**Review by mattsse — APPROVED** [review]
lgtm

File: anvil/src/eth/backend/mem/in_memory_db.rs [hunk 6]
```diff
@@ -46,3 +101,121 @@ impl Db for MemDb {
...
+                code_hash: KECCAK_EMPTY,
+                code: Some(contract_code.clone()),
+                nonce: 1234,
+            },
+        );
+
+        db.set_storage_at(test_addr, "0x1234567".into(), "0x1".into());
+        db.set_storage_at(test_addr, "0x1234568".into(), "0x2".into());
+
+        let mut new_state = SerializableState::new();
```
Comment (mattsse) [inline anvil/src/eth/backend/mem/in_memory_db.rs:183]:
```suggestion
        let mut new_state = SerializableState::default();
```

**Comment by tynes** [comment]
> > > so I g uess my concern is that the dump is kind of useless if it only has the code_hash instead of the actual code
> > 
> > 
> > The example I showed above was the result of `debug_dumpBlock` for an account with no code, the `code` key isn't serialized in the json. When an account does have code, it will have a `code` key, see here: https://github.com/ethereum/go-ethereum/blob/2697e44d819377e39a781e5ab9f1814426b4b0f0/core/state/dump.go#L56
> 
> I see now that the code hash there is the hash of the empty buffer, which makes sense now. So the code of a contract can be guaranteed to be set if it is has code, and I assume same goes for storage.
> 
> Do you happen to have a small sample dump handy? If not, I will try and get a dump from my local eth node tomorrow or something

Here is a link to a dump that came from a local devnet of optimism bedrock - https://gist.github.com/tynes/0cae7ac93e266cd92ad5aa4c37ec95db

**Review by mattsse — APPROVED** [review]

**Comment by dbeal-eth** [comment]
> lgtm
> 
> good to merge @dbeal-eth ?
> 
> it'd be great if you could add them to the anvil reference in https://github.com/foundry-rs/book as well

yes, lets go ahead and merge!

compile issue fixed

book PR https://github.com/foundry-rs/book/pull/447



> > > > so I g uess my concern is that the dump is kind of useless if it only has the code_hash instead of the actual code
> > > 
> > > 
> > > The example I showed above was the result of `debug_dumpBlock` for an account with no code, the `code` key isn't serialized in the json. When an account does have code, it will have a `code` key, see here: https://github.com/ethereum/go-ethereum/blob/2697e44d819377e39a781e5ab9f1814426b4b0f0/core/state/dump.go#L56
> > 
> > 
> > I see now that the code hash there is the hash of the empty buffer, which makes sense now. So the code of a contract can be guaranteed to be set if it is has code, and I assume same goes for storage.
> > Do you happen to have a small sample dump handy? If not, I will try and get a dump from my local eth node tomorrow or something
> 
> Here is a link to a dump that came from a local devnet of optimism bedrock - https://gist.github.com/tynes/0cae7ac93e266cd92ad5aa4c37ec95db

will test running the import with this later and any change is needed will probably submit a follow up PR.

**Comment by gakonst** [comment]
Integration test failure unrelated, fixed in #2286

**Comment by ngotchac** [comment]
Quick question: why leave forked nodes out of this? IMO it would be great to have this for forked node, so that we can easily save, share and restore a forked network with modified state. Maybe I'm missing something though :thinking:

**Comment by dbeal-eth** [comment]
> Quick question: why leave forked nodes out of this? IMO it would be great to have this for forked node, so that we can easily save, share and restore a forked network with modified state. Maybe I'm missing something though thinking

while it is technically possible to save a forked state, I thought I would leave this change to a separate PR because it requires some special handling (ex. in order for the forked state to be guaranteed consistent, it must be forked from exactly the same network and the same block, except when new contracts are being merged into the network, where its not necessarily necessary). The primary inspiration/use case stated above only currently requires local node dump/reload

**Pull request merged**